### PR TITLE
Add code to make Berserk compile on Apple Silicon out of the box.

### PR DESCRIPTION
**Pull request by star26bsd**
Not optmized yet, the arm64 binary is significantly slower than the x86_64 code.

### CODE REVIEW

File: src/makefile [hunk 1]
```diff
@@ -20,13 +20,22 @@ ifeq ($(shell echo "test"), "test")
 	CFLAGS += -static
 endif
 
+# Detecting Apple Silicon (ARM64)
+UNAME := $(shell uname -m)
+ifeq ($(UNAME), arm64)
+    ARCH = arm64
+    CC = clang
```
Comment (jhonnold) [inline src/makefile:27]: Should we be forcing a specific CC like this?

File: src/nn/accumulator.c [hunk 2]
```diff
@@ -16,7 +16,10 @@
 
 #include "accumulator.h"
 
+#if defined(__x86_64__) || defined(_M_X64)
 #include <immintrin.h>
+#endif
+
```
Comment (jhonnold) [inline src/nn/accumulator.c:22]: I actually don't think I use anything from immintrin in this file, so this include can just be removed.

File: src/nn/accumulator.h [hunk 3]
```diff
@@ -17,7 +17,9 @@
 #ifndef ACCUMULATOR_H
 #define ACCUMULATOR_H
 
+#if defined(__x86_64__) || defined(_M_X64)
```
Comment (jhonnold) [inline src/nn/accumulator.h:20]: I actually would prefer if we used the same `defined()` that are used below as that is really what's driving the include being added. I wouldn't mind it being copied into every block either that specifically requires it.

File: src/nn/evaluate.c [hunk 4]
```diff
@@ -16,7 +16,10 @@
 
 #include "evaluate.h"
 
+#if defined(__x86_64__) || defined(_M_X64)
 #include <immintrin.h>
+#endif
```
Comment (jhonnold) [inline src/nn/evaluate.c:21]: Similar to above (but ignore the copy into every place part), I would use the same `defined()` logic that actually causes me to require this include.

**Review by jhonnold — CHANGES_REQUESTED** [review]

**Comment by jhonnold** [comment]
> Not optmized yet, the arm64 binary is significantly slower than the x86_64 code.

Probably have to write native neon or do something like https://github.com/DLTcollab/sse2neon.

Would probably prefer native neon in the long run, though don't have an ARM box to verify with.

File: src/makefile (hunk 1 above)
Comment (star26bsd) [reply src/makefile:27]: No. I guess that is a left-over from being tired using 'make pgo CC=clang' :)

File: src/nn/evaluate.c (hunk 4 above)
Comment (star26bsd) [reply src/nn/evaluate.c:21]: I am not sure I understand correctly. Placing the `defined()` in the proper places would mean to copy it all over indeed, as with accumulator.h file.

Or do you rather prefer:

`#if defined(__AVX512F__) || defined(__AVX512BW__) || defined(__AVX2__) || defined(__SSE2__)`
`#include <immintrin.h> `
`#endif`

**Comment by star26bsd** [comment]
> Probably have to write native neon or do something like https://github.com/DLTcollab/sse2neon.
> Would probably prefer native neon in the long run, though don't have an ARM box to verify with.

A native NEON approach seems reasonable, especially as some intrinsics are rather simple to port. Here's an example for `InputReLu()`

```
+#include <arm_neon.h>
+INLINE void InputReLU(int8_t* outputs, Accumulator* acc, const int stm) {
+    const size_t WIDTH = sizeof(int16x8_t) / sizeof(acc_t);
+    const size_t CHUNKS = N_HIDDEN / WIDTH;
+    const int views[2] = {stm, !stm};
+
+    for (int v = 0; v < 2; v++) {^M
+        const int16x8_t* in = (int16x8_t*) acc->values[views[v]];
+        int8x8_t* out = (int8x8_t*) &outputs[N_HIDDEN * v];
+
+        for (size_t i = 0; i < CHUNKS; i++) {
+            int16x8_t s = vshrq_n_s16(in[i], 5);
+            int8x8_t packed = vqmovn_s16(s);
+            int8x8_t maxed = vmax_s8(packed, vdup_n_s8(0));
+            out[i] = maxed;
+        }
+    }
+}
```


(whereas others like `_mm512_maddubs_epi16` seem to require more work). I am completely out of my depths here, but I am happy to help testing and compiling. I even don't know, if a performant ARM version is even a goal of yours :) 

Cheers
Stephan

**Review by jhonnold — APPROVED** [review]

**Pull request merged**